cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java

```diff
@@ -0,0 +1,252 @@
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.InstanceId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.NamespacedEntityId;
+import co.cask.cdap.proto.id.ProfileId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.runtime.spi.profile.ProfileStatus;
+import com.google.inject.Inject;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This class is to manage preference related functions. It will wrap the {@link PreferencesDataset} operation
+ * in transaction in each method
+ */
+public class PreferencesService {
+
+  private final DatasetFramework datasetFramework;
+  private final Transactional transactional;
+  private final AdminEventPublisher adminEventPublisher;
+  private final MultiThreadMessagingContext messagingContext;
+
+  @Inject
+  public PreferencesService(DatasetFramework datasetFramework, TransactionSystemClient txClient,
+                            AdminEventPublisher adminEventPublisher,
+                            MessagingService messagingService) {
+    this.datasetFramework = datasetFramework;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(new SystemDatasetInstantiator(datasetFramework),
+        txClient, NamespaceId.SYSTEM,
+        Collections.emptyMap(), null, null, messagingContext)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.adminEventPublisher = adminEventPublisher;
+  }
+
+  private Map<String, String> getConfigProperties(EntityId entityId) {
+    return Transactionals.execute(transactional, context -> {
+      return PreferencesDataset.get(context, datasetFramework).getPreferences(entityId);
+    });
+  }
+
+  private Map<String, String> getConfigResolvedProperties(EntityId entityId) {
+    return Transactionals.execute(transactional, context -> {
+      return PreferencesDataset.get(context, datasetFramework).getResolvedPreferences(entityId);
+    });
+  }
+
+  /**
+   * Validate the profile status is enabled and set the preferences in same transaction
+   */
+  private void setConfig(EntityId entityId,
+                         Map<String, String> propertyMap) throws NotFoundException, ProfileConflictException {
+    Transactionals.execute(transactional, context -> {
+      ProfileDataset profileDataset = ProfileDataset.get(context, datasetFramework);
+      PreferencesDataset preferencesDataset = PreferencesDataset.get(context, datasetFramework);
+
+      // need to get old property and check if it contains profile information
+      Map<String, String> oldProperties = preferencesDataset.getPreferences(entityId);
+      preferencesDataset.setPreferences(entityId, propertyMap);
```

Comment:
should set this after the profile check. It will avoid additional writes and a tx rollback in the case that the profile is disabled.

Reply:
Fixed